Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp

```diff
@@ -139,34 +139,28 @@ GStreamerAudioEncoder::~GStreamerAudioEncoder()
     close();
 }
 
-void GStreamerAudioEncoder::encode(RawFrame&& frame, EncodeCallback&& callback)
+Ref<AudioEncoder::EncodePromise> GStreamerAudioEncoder::encode(RawFrame&& frame)
 {
-    gstEncoderWorkQueue().dispatch([frame = WTFMove(frame), encoder = m_internalEncoder, callback = WTFMove(callback)]() mutable {
+    return invokeAsync(gstEncoderWorkQueue(), [frame = WTFMove(frame), encoder = m_internalEncoder]() mutable {
         auto result = encoder->encode(WTFMove(frame));
         if (encoder->isClosed())
-            return;
-
-        String resultString;
-        if (result)
-            encoder->harness()->processOutputSamples();
-        else
-            resultString = "Encoding failed"_s;
+            return EncodePromise::reject("Empty frame"_s);
 
-        encoder->postTask([weakEncoder = ThreadSafeWeakPtr { encoder.get() }, result = WTFMove(resultString), callback = WTFMove(callback)]() mutable {
-            auto encoder = weakEncoder.get();
-            if (!encoder || encoder->isClosed())
-                return;
+        if (!result)
+            return EncodePromise::createAndReject("Encoding failed"_s);
 
-            callback(WTFMove(result));
-        });
+        return EncodePromise::createAndResolve();
     });
 }
 
-void GStreamerAudioEncoder::flush(Function<void()>&& callback)
+Ref<GenericPromise>> GStreamerAudioEncoder::flush()
 {
-    gstEncoderWorkQueue().dispatch([encoder = m_internalEncoder, callback = WTFMove(callback)]() mutable {
-        encoder->flush(WTFMove(callback));
+    return invokeAsync(gstEncoderWorkQueue(), [encoder = m_internalEncoder] {
+        encoder->flush();
```

Comment:
could make the encoder return a promise too and chain them together

Reply:
We would do that if flush was async, but it is not.

Reply:
it was async before. could have kept it async and do :
```
return invokeAsync(gstEncoderWorkQueue(), [encoder = m_internalEncoder] {
        return encoder->flush();
});
```